### ISSUE TITLE
binary encapsulation, not separate protocol

### ISSUE THREAD

**Issue by dormando** [memcached/memcached#490]
With the mget/mset/mdelete work (https://github.com/memcached/memcached/pull/484) and simple authentication tokens (https://github.com/memcached/memcached/pull/489), I've been revisiting the idea of the binary protocol.

The m* commands should be able to do anything the binary protocol can do, plus new extensions, plus future extensions. By comparison, binprot has a heavy default header (24+ bytes), and a 1 byte "extlen" header. Extensions are supposed to be hard coded per command type and fit within the extlen. Over the years I've come to want more of a header-type system. mget/mset demonstrate this: a handful of optional flags that expand/contract the functionality of the command, with overlap between commands. If in the future I want to add a flag to existing commands, it gets added as a flag, rather than an entirely separate command (or set of commands!) in the protocol header.

So at this point, what's lost with binprot that ascii can't do?

1. The opaque/key return for better protocol synchronization/error handling. I'm trying to ensure this works for the m* commands.
2. Binary encoded keys (though many clients never supported this!)
3. 16-bit keys (the server never supported this)
4. Easier parsing due to the length in the header.

I don't want 16bit keys (and nobody asks for them), 1) is being dealt with, so lets focus on 2) and 4). Most people don't need this. I'd say almost everyone doesn't, but a handful of power users will want to min-max their clusters. I'd like to give them the option to do this, without losing too much in the process.

Binary encoding keys is a mixed blessing. For example, if we use a 64bit encoding for an ID number, short numbers end up using more space:

```
num      |     size (bytes)   |  size (string bytes)
1              8                 1
5000           8                 4
65535          8                 5
4b             8                 10
```

... and variable length encoding schemes make clients and servers harder to write. You'll get minimum wire bytes for a loss of flexibility (and some performance).

However, there are some schemes which would improve byte count on the server and greatly reduce string size, useful for extstore users or users with small data values in general.

```
example keys
system:subsystem:subsubsystem:userid:1289427389572351895
{version}:subsystem:subsubsystem:etc:usertoken:ZG9vdCBkb290IGRvb3QgZG9vdCBkb290
```

One option for binpacking keys, which can be at least somewhat easy to parse is to create lookup tables for prefixes. IE: in an extreme case, if there are only 16 possible combinations of "system:subsystem:etc", you can pack the entire prefix down to one byte, though I'd personally at least do 2 bytes :) similarly a version could be a single byte. Then the base64 encoded token id turns into normal binary; whether it was 64bit, 128bit, or a 256bit random token.

Minimal bytes on the wire, fewer cpu cycles spent hashing, far fewer bytes stored, etc.

It's much harder on end users. Cost metrics might make sense for some scenarios, but again I'd urge people to not dive for this unless absolutely necessary.

4), parsing, is an interesting case. For binprot we know to keep asking the network for more data until the header is satisfied. For ascii we need to constantly scan the input buffer for newlines, then tokenize, then wait for more bytes, etc. Further, when we scan for newlines, we have to scan past a potentially long key. Overall it doesn't add up to much CPU time (syscalls are still a grand bulk of overhead), but this can change with pipelining, aggregating clients, and so on.

---

Proposal: solve binary keys and improve parsing speed via a simple wrapper. Our "magic byte" (< 127 for ASCII prot, 0x80/0x81 for binprot) serves as a version byte. We can take four more positions in here (req/resp) to add two more states:

```
short state:
  byte 1: magic byte (0x82 or so)
  byte 2: command length
  byte 3: key length
  command bytes\r\n
  key bytes

long state:
  byte 1: magic byte (0x84 or so)
  byte 2-5: payload length (includes the length of bytes 6,7 and their data!)
  byte 6: command length
  byte 7: key length
```

With this change, a router can easily know how many more bytes to read before parsing a command: first three bytes, then X+Y bytes, or first 5 bytes, then N bytes. The ascii commands are then encapsulated as normal, except the key is removed (or zeroed out) from its normal position in the argument list. Another benefit is the protocol is still legible with tcp dump; a couple header bytes and then an english-ish command. Payloads have always been arbitrary binary (followed by \r\n...)

It's tempting to add one more state for a smaller payload length, but I'd rather simplify it since these would handle SET cases which should be relatively rare.

In summary, you would add 3 or 7 bytes "on the wire" in order to cut N bytes off your keys or improve M% CPU performance for clients/servers/routers. All ASCII current and future commands work under this encapsulation. I've been lazy to keep binprot up to parity with ascii's tuning commands, and this would help in allowing us to eventually deprecate and isolate/remove binprot in the long term.

**Comment by dormando** [comment]
meh. think this is only useful if routers are talking to each other. clients would never do this.

in meta I ended up adding base64 keys to get binary encoding.
